metrics/collector.go

```diff
@@ -0,0 +1,180 @@
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/cybozu-go/log"
+	"github.com/cybozu-go/sabakan/v2"
+	"github.com/cybozu-go/well"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+const (
+	namespace     = "sabakan"
+	scrapeTimeout = time.Second * 10
+)
+
+type logger struct{}
+
+func (l logger) Println(v ...interface{}) {
+	log.Error(fmt.Sprint(v...), nil)
+}
+
+// Metric represents collectors and updater of metric.
+type Metric struct {
+	collectors []prometheus.Collector
+	updater    func(context.Context, *sabakan.Model) error
+}
+
+// Collector is a metrics collector for Sabakan.
+type Collector struct {
+	metrics map[string]Metric
+	model   *sabakan.Model
+}
+
+// NewCollector returns a new Collector.
+func NewCollector(model *sabakan.Model) *Collector {
+	return &Collector{
+		metrics: map[string]Metric{
+			"machine_status": {
+				collectors: []prometheus.Collector{MachineStatus},
+				updater:    updateMachineStatus,
+			},
+			"api_request_count": {
+				collectors: []prometheus.Collector{APIRequestTotal},
+				updater:    updateNop,
+			},
+			"assets_total": {
+				collectors: []prometheus.Collector{AssetsBytesTotal, AssetsItemsTotal},
+				updater:    updateAssetMetrics,
+			},
+			"images_total": {
+				collectors: []prometheus.Collector{ImagesBytesTotal, ImagesItemsTotal},
+				updater:    updateImageMetrics,
+			},
+		},
+		model: model,
+	}
+}
+
+// GetHandler return http.Handler for prometheus metrics
+func GetHandler(collector *Collector) http.Handler {
+	registry := prometheus.NewRegistry()
+	registry.MustRegister(collector)
+
+	handler := promhttp.HandlerFor(registry,
+		promhttp.HandlerOpts{
+			ErrorLog:      logger{},
+			ErrorHandling: promhttp.ContinueOnError,
+		})
+
+	return handler
+}
+
+// Describe implements prometheus.Collector.Describe().
+func (c Collector) Describe(ch chan<- *prometheus.Desc) {
+	for _, metric := range c.metrics {
+		for _, col := range metric.collectors {
+			col.Describe(ch)
+		}
+	}
+}
+
+// Collect implements prometheus.Collector.Collect().
+func (c Collector) Collect(ch chan<- prometheus.Metric) {
+	ctx, cancel := context.WithTimeout(context.Background(), scrapeTimeout)
+	defer cancel()
+	env := well.NewEnvironment(ctx)
+	for key, metric := range c.metrics {
+		key, metric := key, metric
+		env.Go(func(ctx context.Context) error {
+			err := metric.updater(ctx, c.model)
+			if err != nil {
+				log.Warn("unable to update metrics", map[string]interface{}{
+					"name":      key,
+					log.FnError: err,
+				})
+				// return nil not to disturb collection of other metrics
+				return nil
+			}
+
+			for _, col := range metric.collectors {
+				col.Collect(ch)
+			}
+			return nil
+		})
+	}
+	env.Stop()
+	env.Wait()
```

Comment:
`well.Environment` might be an overkill.
`sync.WaitGroup` is just enough because there are no error handling code.